File: android-app/src/main/java/fr/cph/chicago/core/fragment/CtaMapFragment.java
package fr.cph.chicago.core.fragment;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.chrisbanes.photoview.PhotoView;

import butterknife.BindView;
import fr.cph.chicago.R;
import fr.cph.chicago.core.App;
import fr.cph.chicago.util.Util;

@SuppressWarnings("WeakerAccess")
public class CtaMapFragment extends AbstractFragment {

    @BindView(R.id.cta_map)
    PhotoView ctaMap;
    private Bitmap bitmapCache;

    private final Util util;

    public CtaMapFragment() {
        util = Util.INSTANCE;
    }

    @NonNull
    public static CtaMapFragment newInstance(final int sectionNumber) {
        return (CtaMapFragment) fragmentWithBundle(new CtaMapFragment(), sectionNumber);
    }

    @Override
    public final void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        util.trackScreen(getString(R.string.analytics_cta_map_fragment));
    }

    @Override
    public final View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_cta_map, container, false);
        if (!activity.isFinishing()) {
            setBinder(rootView);
            loadBitmap(ctaMap);
        }
        return rootView;
    }

    public void loadBitmap(final PhotoView imageView) {
        if (bitmapCache != null) {
            imageView.setImageBitmap(bitmapCache);
        } else {
            final BitmapWorkerTask task = new BitmapWorkerTask(imageView);
            task.execute();
        }
    }

    private class BitmapWorkerTask extends AsyncTask<Void, Void, Bitmap> {
        private final PhotoView imageView;

        BitmapWorkerTask(final PhotoView imageView) {
            this.imageView = imageView;
        }

        @Override
        protected Bitmap doInBackground(final Void... params) {
            return BitmapFactory.decodeResource(getResources(), R.drawable.ctamap);
        }

        @Override
        protected final void onPostExecute(final Bitmap bitmap) {
            CtaMapFragment.this.bitmapCache = bitmap;
            imageView.setImageBitmap(bitmap);
        }
    }
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/fragment/CtaMapFragment.kt
/**
 * Copyright 2021 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.fragment

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import fr.cph.chicago.R
import fr.cph.chicago.databinding.FragmentCtaMapBinding
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import timber.log.Timber

class CtaMapFragment : Fragment() {

    companion object {
        fun newInstance(sectionNumber: Int): CtaMapFragment {
            return fragmentWithBundle(CtaMapFragment(), sectionNumber) as CtaMapFragment
        }
    }

    private var bitmapCache: Bitmap? = null

    private var _binding: FragmentCtaMapBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentCtaMapBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loadBitmap()
    }

    @SuppressLint("CheckResult")
    private fun loadBitmap() {
        Observable.fromCallable {
            if (bitmapCache != null) bitmapCache
            else BitmapFactory.decodeResource(resources, R.drawable.ctamap)
        }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { bitmap ->
                    this@CtaMapFragment.bitmapCache = bitmap
                    binding.ctaMap.setImageBitmap(bitmap)
                },
                { error -> Timber.e(error) })
    }
}
